src/vmm/src/devices/virtio/vhost_user.rs

```diff
@@ -0,0 +1,264 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Portions Copyright 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::os::fd::AsRawFd;
+use std::os::unix::net::UnixStream;
+
+use thiserror::Error;
+use utils::eventfd::EventFd;
+use vhost::vhost_user::message::*;
+use vhost::vhost_user::{Master, VhostUserMaster};
+use vhost::{Error as VhostError, VhostBackend, VhostUserMemoryRegionInfo, VringConfigData};
+use vm_memory::{Address, Error as MmapError, GuestMemory, GuestMemoryError, GuestMemoryRegion};
+
+use crate::devices::virtio::device::IrqTrigger;
+use crate::devices::virtio::queue::Queue;
+use crate::vstate::memory::GuestMemoryMmap;
+
+/// vhost-user error.
+#[derive(Error, Debug, displaydoc::Display)]
+pub enum VhostUserError {
+    /// Invalid available address
+    AvailAddress(GuestMemoryError),
+    /// Failed to connect to UDS Unix stream: {0}
+    Connect(#[from] std::io::Error),
+    /// Invalid descriptor table address
+    DescriptorTableAddress(GuestMemoryError),
+    /// Get features failed: {0}
+    VhostUserGetFeatures(VhostError),
+    /// Get protocol features failed: {0}
+    VhostUserGetProtocolFeatures(VhostError),
+    /// Set owner failed: {0}
+    VhostUserSetOwner(VhostError),
+    /// Set features failed: {0}
+    VhostUserSetFeatures(VhostError),
+    /// Set protocol features failed: {0}
+    VhostUserSetProtocolFeatures(VhostError),
+    /// Set mem table failed: {0}
+    VhostUserSetMemTable(VhostError),
+    /// Set vring num failed: {0}
+    VhostUserSetVringNum(VhostError),
+    /// Set vring addr failed: {0}
+    VhostUserSetVringAddr(VhostError),
+    /// Set vring base failed: {0}
+    VhostUserSetVringBase(VhostError),
+    /// Set vring call failed: {0}
+    VhostUserSetVringCall(VhostError),
+    /// Set vring kick failed: {0}
+    VhostUserSetVringKick(VhostError),
+    /// Set vring enable failed: {0}
+    VhostUserSetVringEnable(VhostError),
+    /// Failed to read vhost eventfd: {0}
+    VhostUserMemoryRegion(MmapError),
+    /// Invalid used address
+    UsedAddress(GuestMemoryError),
+}
+
+/// vhost-user socket handle
+#[derive(Clone)]
+pub struct VhostUserHandle {
```

Comment:
discussed offline, we'd like to try making it a trait to be able to mock the implementation for unit testing.